### ISSUE TITLE
[Performance]  Very slow predictions Onnx vs Tensorflow in python

### ISSUE THREAD

**Issue by roushrsh** [microsoft/onnxruntime#15431]
### Describe the issue

My specific issue I've been trying 2 weeks now to solve with different libraries is speed. 

Prediction speed on 500 samples:

Microsoft.ML
GPU: 
 var transformedData = multiPipeline.Fit(dataviewTest).Transform(dataviewTest);  var column = transformedData.GetColumn<float[]>("dense_55").ToArray();                              (250ms)

Python:
Model.predict(data,verbose=False) 40ms
mode(data,training=False) 0.3ms

Tensorflow.NET:
model.predict((xx1), batch_size: 128); 300ms (GPU)
model.predict((xx1), batch_size: 128); 15000ms (CPU)

GPU 4090. 

Not sure what else to try, or what I'm missing to get the predictions closer to python speed.

It's a 800,000 param Yolo model attached. Perhaps it's the conversion to onnx, but I can't test it with tf2.8, batchnorm 15 and squeeze 13 of this model on python with onnx.

I've attached the onnx model (acquired from tfToOnnx), and the original h5 model.  I can see the GPU jump to 60% utilization when it's running. Couldn't get OnnxRuntime.DirectML working to give that a tru over .GPU.

I also tried quantizing it to float16 but the onnx convertor shows it doesn't yet support ft16, so I gave up on that.

### To reproduce

Code for onnx on c#:

  public class OnnxInput
        {
            [VectorType(4319, 1)]
            [ColumnName("input_179")]
            public float[] Image1 { get; set; }

            [VectorType(4319, 1)]
            [ColumnName("input_180")]
            public float[] Image2 { get; set; }

            [VectorType(4319, 1)]
            [ColumnName("input_181")]
            public float[] Image3 { get; set; }

            [VectorType(4319, 1)]
            [ColumnName("input_182")]
            public float[] Image4 { get; set; }

            [ColumnName("input_183")]
            public float CH{ get; set; }

            [ColumnName("input_184")]
            public float PE { get; set; }

            [ColumnName("input_185")]
            public float OI { get; set; }

            [ColumnName("input_186")]
            public float IO { get; set; }

            [ColumnName("input_187")]
            public float OR { get; set; }
        }

        public class OnnxOutput
        {
            [ColumnName("dense_55")]
            public float[] output{ get; set; }
        }

float[] image1 = new float[] { 0.0f, ...} ;// 4319 times, just make a random input for testing

               var testInput = new OnnxInput
               {
                   Image1 = image1,
                   Image2 = image1,
                   Image3 = image1,
                   Image4 = image1,
                   CH= 3f,
                   PE= 1.85F,
                   OI= 1f,
                   OR= 1f,
                   IO= 0f

               };
OnnxInput[] fiveHundredExamples = new OnnxInput[]
               {
testInput,testInput... //(500times for a quick check)
}

            var multiPipeline = mlContext.Transforms.ApplyOnnxModel(modelFile: ONNX_MODEL_PATH, outputColumnNames: new[] { "dense_55" }, inputColumnNames: new[] { "input_179", "input_180", "input_181", "input_182", "input_183", "input_184", "input_185", "input_186", "input_187" }, gpuDeviceId: 0);

            IDataView dataviewTest = mlContext.Data.LoadFromEnumerable<OnnxInput>(fiveHundredExamples );
            var transformedData = multiPipeline.Fit(dataviewTest).Transform(dataviewTest);
            var column = transformedData.GetColumn<float[]>("dense_55").ToArray();

            watchz = new System.Diagnostics.Stopwatch();
            watchz.Start();
            var transformedData2 = multiPipeline.Fit(dataviewTest).Transform(dataviewTest);
            var columnz2 = transformedData2.GetColumn<float[]>("dense_55").ToArray();
            watchz.Stop();
            Console.WriteLine($"Execution Time: {watchz.ElapsedMilliseconds} ms");

### Platform

Windows

### OS Version

Windows 10

### ONNX Runtime Installation

Built from Source

### ONNX Runtime Version or Commit ID

1.14.1 GPU

### ONNX Runtime API

C#

### Architecture

X64

### Execution Provider

CUDA

### Execution Provider Library Version

Cuda 11.2

### Model File

[unTrainTest.zip](https://github.com/microsoft/onnxruntime/files/11184926/unTrainTest.zip)

### Is this a quantized model?

No

**Comment by roushrsh** [comment]
Update, testing the onnx on python for the same model and it's exactly the same, slightly faster even.... just need to figure out what is going on in c# and I'm set.

**Comment by roushrsh** [comment]
Update 2: Keras.net predictonbatch is 50ms, but it's so old (and not supported anymore) that everything has to be converted to NDarrays, which makes the whole thing 300ms anyways. Shucks.

**Comment by roushrsh** [comment]
Update 3: CPU is running faster than GPU. Something definitely is wrong, even though I see a spike in GPU use.

**Comment by roushrsh** [comment]
If anyone could just run the model with an empty input and see if it's slower on gpu for them or not as well it would help a lot.

**Comment by roushrsh** [comment]
Update 4:
I ran the model on a docker and another PC entirely. Problem remains. The model is slow to run on C#, perhaps a fundamental issue with how the code is.

**Comment by pauldog** [comment]
Hello, from my experience there is a known issue that changing the size of the inputs of a model will drastically affect the inference times. i.e. for the best times in DirectML you need to keep the size of the input and batch number the same for all inferences. Do you think this might be related to your timings? (Also, I think on average DirectML is usually a bit slower than CUDA)

**Comment by roushrsh** [comment]
Thanks for the reply. I'm using GPU, not DirectML. 
Model input is the same. How are you choosing batch number when using .Transform for multi-prediction?
I replied to you in the chat as well.

**Comment by nums11** [comment]
Closing as stale.
